### ISSUE TITLE
documentation: puma with phased-restart

### ISSUE THREAD

**Issue by skyeagle** [chanks/que#89]
Hello, @chanks !

In accordance with [this](https://github.com/puma/puma/issues/656#issuecomment-74598638) comment, puma configured with `phased-restart` doesn't require to use `on_worker_boot` hook. I had a success only with following setup:

```ruby
# config/application.rb
Que.worker_count = 0
Que.mode = :off
```

```ruby
# config/environments/development.rb
ActiveSupport.on_load(:after_initialize) do
  Que.mode = :async
  Que::Worker.workers.push Que::Worker.new 'default'
end
```

It will be great if you fix the documentation. Thank you.

**Comment by chanks** [comment]
I'm not sure how phased-restart works in Puma, but I think it's probably a bad idea to manually start workers and push them to the array like you're doing. Does the worker_count= approach not work there?

**Comment by skyeagle** [comment]
If you need zero downtime restart then you must use phased-restart. I filled workers in this way because it is easy to manage different queues. So, workers_count = will work too. My first implementation didn't deal with rake tasks, here the improved one:

```ruby
# config/application.rb
Que.worker_count = 0
Que.mode = :off
```

```ruby
# config/puma/development.rb
workers 2
on_worker_boot do
  ActiveSupport.on_load(:after_initialize) do
    Que.mode = :async
    Que::Worker.workers.push Que::Worker.new 'default'
  end
end
```

**Comment by j4zz** [comment]
I’m sorry if I’m hijacking the issue, but I experienced a problem where Que workers were not running on Heroku and the configuration provided by @skyeagle solved it for me so it might be related.

I’m confident that’s a configuration issue on my side and I lack understanding of Que, but when following [the forking servers documentation](https://github.com/chanks/que/blob/master/docs/advanced_setup.md#forking-servers) I was unsure what my mistake could be.

@chanks you mentioned using `worker_count=`? Did you mean like in the following example? Because the following does not correctly spawn active workers for me, whereas `Que::Worker.workers.push Que::Worker.new 'default'` like mentioned by @skyeagle does.


```ruby
# config/application.rb
Que.worker_count = 0
Que.mode = :off
```

```ruby
# config/unicorn.rb
after_fork do |server, worker|
  ActiveRecord::Base.establish_connection

  Que.mode = :async
  Que.worker_count = 4
end
```

**Comment by chanks** [comment]
I'd expect that worker_count setting to work, yes. I'm not sure why it wouldn't, but then I don't really know Unicorn that well.

A couple questions for debugging this:
- The after_fork block is run in the forked worker, right? Not in the parent process?
- What are the values of Que.mode and Que.worker_count in the after_fork block before they are set?

I suspect that the settings in application.rb are not taking effect early enough in the bootup process, so there are four workers running during the fork. Then the fork kills their threads, but since the four (dead) workers are still in the array, setting worker_count to 4 doesn't do anything. Someone who knows the Rails boot process better might be able to tell us a better place for that code. In the meantime, another workaround that might work is setting Que.worker_count = 0 then Que.worker_count = 4 in that after_fork block.

**Comment by j4zz** [comment]
> The after_fork block is run in the forked worker, right? Not in the parent process?

It states in [the documentation](http://unicorn.bogomips.org/Unicorn/Configurator.html#method-i-after_fork) that `This block [after_fork] will be called by the worker after forking.` For contrast, `before_fork` is listed in the documentation as being called by the master process.

> What are the values of Que.mode and Que.worker_count in the after_fork block before they are set?

They seem to be gathered from `application.rb` and are `Que.mode: off` and `Que.worker_count: 0` from the following snippet:

```ruby
# config/unicorn.rb
after_fork do |server, worker|
  ActiveRecord::Base.establish_connection
  puts "Que.mode: #{Que.mode}"
  puts "Que.worker_count: #{Que.worker_count}"
  Que.mode = :async
  Que.worker_count = 4
end
```

> In the meantime, another workaround that might work is setting Que.worker_count = 0 then Que.worker_count = 4 in that after_fork block.

I’ve given the proposed workaround a try on Heroku and it doesn’t work for me.

```ruby
# config/unicorn.rb
after_fork do |server, worker|
  ActiveRecord::Base.establish_connection
  Que.mode = :async
  Que.worker_count = 0
  Que.worker_count = 4
end
```

**Comment by chanks** [comment]
That's... frustrating. There goes that theory.

I'll try to reproduce this when I have time (probably during an open-source Friday at work), but I'm not sure when that'll be.

**Comment by j4zz** [comment]
No pressure on my side. If you've got something to test, I can give it a try on Heroku. From the testing so far, using foreman seems to be enough to simulate Heroku behavior, though.

**Comment by chanks** [comment]
Hey @j4zz - so, enough people have had problems with the autostarting not working as it should that I'm trying to replace it entirely with Sidekiq or Hutch-style CLI, rather than forcing people to do battle with their forking webservers to get things running properly. It's in #112 - hopefully if it works out we won't be having these sorts of problems anymore.

**Comment by chanks** [comment]
The autostarting fix has shipped in 0.11.1 now, so I'm closing this. Thanks!
